src/StableBuffer.mo

```diff
@@ -86,88 +89,95 @@ module {
+  public func clone<X>(buffer : StableBuffer<X>) : StableBuffer<X> {
     let c = initPresized<X>(buffer.elems.size());
-    var i = 0;
+    var i = vals(buffer);
+
     label l loop {
-      if (i >= buffer.count) break l;
-      add(c, buffer.elems[i]);
-      i += 1;
+      switch (i.next()) {
+        case null break l;
+        case (?x) { add(c, x) };
+      };
     };
-    c
+
+    c;
   };
 
   /// Returns an `Iter` over the elements of this buffer.
-  public func vals<X>(buffer: StableBuffer<X>) : { next : () -> ?X } = object {
+  public func vals<X>(buffer : StableBuffer<X>) : { next : () -> ?X } = object {
     var pos = 0;
     public func next() : ?X {
       if (pos == buffer.count) { null } else {
-        let elem = ?buffer.elems[pos];
+        let elem = buffer.elems[pos];
         pos += 1;
-        elem
-      }
-    }
+        elem;
+      };
+    };
   };
 
   /// Creates a Buffer from an Array
-  public func fromArray<X>(xs: [X]): StableBuffer<X> {
-    let ys: StableBuffer<X> = initPresized(xs.size());
+  public func fromArray<X>(xs : [X]) : StableBuffer<X> {
+    let ys : StableBuffer<X> = initPresized(xs.size());
     for (x in xs.vals()) {
       add(ys, x);
     };
 
-    ys
+    ys;
   };
 
   /// Creates a new array containing this buffer's elements.
-  public func toArray<X>(buffer: StableBuffer<X>) : [X] =
-    // immutable clone of array
-    Prim.Array_tabulate<X>(
-      buffer.count,
-      func(x : Nat) : X { buffer.elems[x] }
-    );
+  public func toArray<X>(buffer : StableBuffer<X>) : [X] =
+  // immutable clone of array
+  Prim.Array_tabulate<X>(
+    buffer.count,
+    func(x : Nat) : X {
+      get(buffer, x);
+    },
+  );
 
   /// Creates a mutable array containing this buffer's elements.
-  public func toVarArray<X>(buffer: StableBuffer<X>) : [var X] {
+  public func toVarArray<X>(buffer : StableBuffer<X>) : [var X] {
     if (buffer.count == 0) { [var] } else {
-      let a = Prim.Array_init<X>(buffer.count, buffer.elems[0]);
+      let a = Prim.Array_init<X>(buffer.count, get(buffer, 0));
       var i = 0;
       label l loop {
         if (i >= buffer.count) break l;
-        a[i] := buffer.elems[i];
+        a[i] := get(buffer, i);
         i += 1;
       };
-      a
-    }
+
+      a;
+    };
   };
 
   /// Gets the `i`-th element of this buffer. Traps if  `i >= count`. Indexing is zero-based.
-  public func get<X>(buffer: StableBuffer<X>, i : Nat) : X {
-    assert(i < buffer.count);
-    buffer.elems[i]
+  public func get<X>(buffer : StableBuffer<X>, i : Nat) : X {
+    switch (buffer.elems[i]) {
+      case (?elem) elem;
+      case (null) Prim.trap("Buffer index out of bounds in get");
+    };
   };
 
   /// Gets the `i`-th element of the buffer as an option. Returns `null` when `i >= count`. Indexing is zero-based.
-  public func getOpt<X>(buffer: StableBuffer<X>, i : Nat) : ?X {
+  public func getOpt<X>(buffer : StableBuffer<X>, i : Nat) : ?X {
```

Comment:
If the buffer is changed internally to have elems of type `?X`, then is this function still needed? (Can just use `get`)

Reply:
I think this function still serves its purpose even with the `?X` type. This function retrieves the value of any index inside or outside the buffer without trapping.